File: app/src/main/java/pl/szczodrzynski/edziennik/login/LoginMobidziennikFragment.java
package pl.szczodrzynski.edziennik.login;

import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.danimahardhika.cafebar.CafeBar;

import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import pl.szczodrzynski.edziennik.App;
import pl.szczodrzynski.edziennik.R;
import pl.szczodrzynski.edziennik.api.AppError;
import pl.szczodrzynski.edziennik.databinding.FragmentLoginMobidziennikBinding;

import static pl.szczodrzynski.edziennik.api.AppError.CODE_ARCHIVED;
import static pl.szczodrzynski.edziennik.api.AppError.CODE_INVALID_LOGIN;
import static pl.szczodrzynski.edziennik.api.AppError.CODE_INVALID_SERVER_ADDRESS;
import static pl.szczodrzynski.edziennik.api.AppError.CODE_OLD_PASSWORD;
import static pl.szczodrzynski.edziennik.datamodels.LoginStore.LOGIN_TYPE_MOBIDZIENNIK;

public class LoginMobidziennikFragment extends Fragment {

    private App app;
    private NavController nav;
    private FragmentLoginMobidziennikBinding b;
    private static final String TAG = "LoginMobidziennik";

    public LoginMobidziennikFragment() { }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        if (getActivity() != null) {
            app = (App) getActivity().getApplicationContext();
            nav = Navigation.findNavController(getActivity(), R.id.nav_host_fragment);
        }
        else {
            return null;
        }
        b = DataBindingUtil.inflate(inflater, R.layout.fragment_login_mobidziennik, container, false);
        return b.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        assert getContext() != null;
        assert getActivity() != null;

        view.postDelayed(() -> {
            AppError error = LoginActivity.error;
            if (error != null) {
                switch (error.errorCode) {
                    case CODE_INVALID_LOGIN:
                        b.loginPasswordLayout.setError(getString(R.string.login_error_incorrect_login_or_password));
                        break;
                    case CODE_OLD_PASSWORD:
                        b.loginPasswordLayout.setError(getString(R.string.login_error_old_password));
                        break;
                    case CODE_ARCHIVED:
                        b.loginUsernameLayout.setError(getString(R.string.sync_error_archived));
                        break;
                    case CODE_INVALID_SERVER_ADDRESS:
                        b.loginServerAddressLayout.setError(getString(R.string.login_error_incorrect_address));
                        break;
                    default:
                        CafeBar.builder(getActivity())
                                .to(b.root)
                                .content(getString(R.string.login_error, error.asReadableString(getActivity())))
                                .autoDismiss(false)
                                .positiveText(R.string.ok)
                                .onPositive(CafeBar::dismiss)
                                .floating(true)
                                .swipeToDismiss(true)
                                .neutralText(R.string.more)
                                .onNeutral(cafeBar -> app.apiEdziennik.guiShowErrorDialog(getActivity(), error, R.string.error_details))
                                .negativeText(R.string.report)
                                .onNegative((cafeBar -> app.apiEdziennik.guiReportError(getActivity(), error, null)))
                                .show();
                        break;
                }
                LoginActivity.error = null;
            }
        }, 100);

        b.helpButton.setOnClickListener((v) -> nav.navigate(R.id.loginMobidziennikHelpFragment, null, LoginActivity.navOptions));
        b.backButton.setOnClickListener((v) -> nav.navigateUp());

        b.loginButton.setOnClickListener((v) -> {
            boolean errors = false;

            b.loginServerAddressLayout.setError(null);
            b.loginUsernameLayout.setError(null);
            b.loginPasswordLayout.setError(null);

            Editable serverNameEditable = b.loginServerAddress.getText();
            Editable usernameEditable = b.loginUsername.getText();
            Editable passwordEditable = b.loginPassword.getText();
            if (serverNameEditable == null || serverNameEditable.length() == 0) {
                b.loginServerAddressLayout.setError(getString(R.string.login_error_no_address));
                errors = true;
            }
            if (usernameEditable == null || usernameEditable.length() == 0) {
                b.loginUsernameLayout.setError(getString(R.string.login_error_no_login));
                errors = true;
            }
            if (passwordEditable == null || passwordEditable.length() == 0) {
                b.loginPasswordLayout.setError(getString(R.string.login_error_no_password));
                errors = true;
            }

            if (errors)
                return;
            errors = false;

            String serverName = serverNameEditable.toString().toLowerCase().replaceAll("(?:http://|www.|mobidziennik\\.pl|wizja\\.net|\\.)", "");
            String username = usernameEditable.toString().toLowerCase();
            String password = passwordEditable.toString();
            b.loginServerAddress.setText(serverName);
            b.loginUsername.setText(username);
            if (!serverName.matches("^[a-z0-9_\\-]+$")) {
                b.loginServerAddressLayout.setError(getString(R.string.login_error_incorrect_address));
                errors = true;
            }
            if (!username.matches("^[a-z0-9_\\-@+.]+$")) {
                b.loginUsernameLayout.setError(getString(R.string.login_error_incorrect_login));
                errors = true;
            }

            if (errors)
                return;
            errors = false;

            Bundle args = new Bundle();
            args.putInt("loginType", LOGIN_TYPE_MOBIDZIENNIK);
            args.putString("serverName", serverName);
            args.putString("username", username);
            args.putString("password", password);
            nav.navigate(R.id.loginProgressFragment, args, LoginActivity.navOptions);
        });
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/modules/login/LoginMobidziennikFragment.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2020-1-3.
 */

package pl.szczodrzynski.edziennik.ui.modules.login

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import pl.szczodrzynski.edziennik.*
import pl.szczodrzynski.edziennik.data.api.*
import pl.szczodrzynski.edziennik.databinding.FragmentLoginMobidziennikBinding
import java.util.*
import kotlin.coroutines.CoroutineContext

class LoginMobidziennikFragment : Fragment(), CoroutineScope {
    companion object {
        private const val TAG = "LoginMobidziennikFragment"
    }

    private lateinit var app: App
    private lateinit var activity: LoginActivity
    private lateinit var b: FragmentLoginMobidziennikBinding
    private val nav by lazy { activity.nav }

    private val job: Job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        activity = (getActivity() as LoginActivity?) ?: return null
        context ?: return null
        app = activity.application as App
        b = FragmentLoginMobidziennikBinding.inflate(inflater)
        return b.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        activity.lastError?.let { error ->
            activity.lastError = null
            startCoroutineTimer(delayMillis = 100) {
                when (error.errorCode) {
                    ERROR_LOGIN_MOBIDZIENNIK_WEB_INVALID_LOGIN ->
                        b.loginPasswordLayout.error = getString(R.string.login_error_incorrect_login_or_password)
                    ERROR_LOGIN_MOBIDZIENNIK_WEB_OLD_PASSWORD ->
                        b.loginPasswordLayout.error = getString(R.string.login_error_old_password)
                    ERROR_LOGIN_MOBIDZIENNIK_WEB_ARCHIVED ->
                        b.loginUsernameLayout.error = getString(R.string.sync_error_archived)
                    ERROR_LOGIN_MOBIDZIENNIK_WEB_INVALID_ADDRESS ->
                        b.loginServerAddressLayout.error = getString(R.string.login_error_incorrect_address)
                }
            }
        }

        b.helpButton.onClick { nav.navigate(R.id.loginMobidziennikHelpFragment, null, LoginActivity.navOptions) }
        b.backButton.onClick { nav.navigateUp() }

        b.loginButton.onClick {
            var errors = false

            b.loginServerAddressLayout.error = null
            b.loginUsernameLayout.error = null
            b.loginPasswordLayout.error = null

            val serverName = b.loginServerAddress.text
                    ?.toString()
                    ?.toLowerCase(Locale.ROOT)
                    ?.replace("(?:http://|www.|mobidziennik\\.pl|wizja\\.net|\\.)".toRegex(), "") ?: ""
            val username = b.loginUsername.text?.toString()?.toLowerCase(Locale.ROOT) ?: ""
            val password = b.loginPassword.text?.toString() ?: ""

            if (serverName.isBlank()) {
                b.loginServerAddressLayout.error = getString(R.string.login_error_no_address)
                errors = true
            }
            if (username.isBlank()) {
                b.loginUsernameLayout.error = getString(R.string.login_error_no_login)
                errors = true
            }
            if (password.isBlank()) {
                b.loginPasswordLayout.error = getString(R.string.login_error_no_password)
                errors = true
            }
            if (errors) return@onClick

            errors = false

            b.loginServerAddress.setText(serverName)
            b.loginUsername.setText(username)
            if (!"^[a-z0-9_\\-]+$".toRegex().matches(serverName)) {
                b.loginServerAddressLayout.error = getString(R.string.login_error_incorrect_address)
                errors = true
            }
            if (!"^[a-z0-9_\\-@+.]+$".toRegex().matches(username)) {
                b.loginUsernameLayout.error = getString(R.string.login_error_incorrect_login)
                errors = true
            }
            if (errors) return@onClick

            val args = Bundle(
                    "loginType" to LOGIN_TYPE_MOBIDZIENNIK,
                    "serverName" to serverName,
                    "username" to username,
                    "password" to password
            )
            nav.navigate(R.id.loginProgressFragment, args, LoginActivity.navOptions)
        }
    }
}